packages/tenant-discovery-nodejs/src/plugin.ts

```diff
@@ -0,0 +1,93 @@
+import { SuperTokensPlugin } from "supertokens-node/types";
+import { createPluginInitFunction } from "@shared/js";
+import {
+  OverrideableTenantFunctionImplementation,
+  SuperTokensPluginTenantDiscoveryPluginConfig,
+  SuperTokensPluginTenantDiscoveryPluginNormalisedConfig,
+} from "./types";
+import { HANDLE_BASE_PATH, PLUGIN_ID, PLUGIN_SDK_VERSION } from "./constants";
+import { POPULAR_EMAIL_DOMAINS } from "./constants";
+import { withRequestHandler } from "@shared/nodejs";
+import { enableDebugLogs, logDebugMessage } from "./logger";
+
+import { getOverrideableTenantFunctionImplementation } from "./recipeImplementation";
+
+export const init = createPluginInitFunction<
+  SuperTokensPlugin,
+  SuperTokensPluginTenantDiscoveryPluginConfig,
+  OverrideableTenantFunctionImplementation,
+  SuperTokensPluginTenantDiscoveryPluginNormalisedConfig
+>(
+  (pluginConfig, implementation): SuperTokensPlugin => {
+    return {
+      id: PLUGIN_ID,
+      compatibleSDKVersions: PLUGIN_SDK_VERSION,
+      init: async (appConfig) => {
+        if (appConfig.debug) {
+          enableDebugLogs();
+        }
+
+        // Ensure that the domain is not a generic one
+        // Go through each domain and check if it is a popular one
+        const restrictedDomains = [...(pluginConfig.restrictedEmailDomains ?? []), ...POPULAR_EMAIL_DOMAINS];
+
+        logDebugMessage(`Updated Restricted domains: ${restrictedDomains}`);
+      },
+      routeHandlers: () => {
+        return {
+          status: "OK",
+          routeHandlers: [
+            {
+              path: `${HANDLE_BASE_PATH}/list`,
+              method: "get",
+              handler: withRequestHandler(async (req, res, session, useContext) => {
+                const tenants = await implementation.getTenants();
```

Comment:
I think we should expose this only as an opt-in thing (returning some kind of error), as it could leak information about the client list of an app.